### PR TITLE
Add static typing support

### PR DESCRIPTION
**Pull request by headtr1ck**
Closes: #1298

Lots of this PR was stolen from https://github.com/Unidata/netcdf4-python/pull/1279

Notice: I intentionally reversed the import behavior of `__init__.py` and `_netCDF4.pyx` in the stubs. This should be more in line in how people are using netCDF4. And while doing so I also removed the "._netCDF4" part of the classname in the reprs.

The current typing is working with python 3.11.
TODO: 
- [x] ~~Make typing pass with python 3.8 (aka. minimum supported version)~~ Won't do

### CODE REVIEW

**Comment by jswhit** [comment]
Thanks!  Looks good - but why did you add a new workflow to run stubtest, instead of adding the extra test to the existing workflows?

**Comment by headtr1ck** [comment]
> why did you add a new workflow to run stubtest, instead of adding the extra test to the existing workflows?

Good question. I just took the changes from the previous PR #1279

After the holidays I can continue to work on this and use the existing workflows.

**Comment by jswhit** [comment]
Looks you like need to add `git submodule update --init --recursive` after the `git clone` in `stubtest.yml`.

**Comment by headtr1ck** [comment]
Hey, I was just very busy the last months.
I can try to get it to work the next days.

And thanks for your initial work, it was very helpful to get started.

**Comment by headtr1ck** [comment]
I have now made Variable a generic class to solve the datatype problem.
Unfortunately, at least for me, it does not seem to be able to resolve the non-np.dtype np.DTypeLikes.
So if you as datatype e.g. `netCDF4.VLType` or `np.int32` it works, but `i4` or `float32` will not resolve to `np.dtype` but to their Literals. I have not been able to solve this problem.

But I think it is already better than before and should also not block merging this PR. We can always fix this in the future as it is a minor problem.

One sad thing, is that we have to return Any for __getitem__ at Dataset/Group, because statically we cannot know if it returns a group or a variable. I don't think there is any solution for that.

**Comment by Woefie** [comment]
> One sad thing, is that we have to return Any for `__getitem__ ` at Dataset/Group, because statically we cannot know if it returns a group or a variable. I don't think there is any solution for that.

Wouldn't a Union work?
`def __getitem__(self, elem: str) -> Union[Group|Variable]: ...`

**Comment by headtr1ck** [comment]
> Wouldn't a Union work?
> `def __getitem__(self, elem: str) -> Union[Group, Variable]: ...`

 Technically yes, but this will result in many false positive errors from mypy on user side.
The recommended solution for that is to return Any and never return Unions.

**Comment by headtr1ck** [comment]
A sidenote:

- I have removed the `_netCDF4` module path from the reprs, I doubt that should affect anyone but one could consider this a breaking change.
- I was adding `always_mask` to private_attrs, because I think it is correct after reading the C code (As well as to type it as read-only property).
- Does anyone have objections on the matter that from a typing perspective the classes live in the __init__ and not in the _netCDF4 module?

Other than that, this is good to go :)

**Comment by jswhit** [comment]
> A sidenote:
> 
> * I have removed the `_netCDF4` module path from the reprs, I doubt that should affect anyone but one could consider this a breaking change.
> * I was adding `always_mask` to private_attrs, because I think it is correct after reading the C code (As well as to type it as read-only property).
> * Does anyone have objections on the matter that from a typing perspective the classes live in the `__init__` and not in the `_netCDF4` module?
> 
> Other than that, this is good to go :)

@headtr1ck please see my questions in the review

**Comment by headtr1ck** [comment]
> @headtr1ck please see my questions in the review

I don't see a review? Are you sure you published it?

**Comment by headtr1ck** [comment]
Don't know if anyone noticed, but `dtype_is_complex` seems to be wrong? It only works if you pass `c8` or `c16`, but not `np.dtype("c8")`.

**Comment by jswhit** [comment]
> > @headtr1ck please see my questions in the review
> 
> I don't see a review? Are you sure you published it?

oops, just submitted it

File: src/netCDF4/__init__.pyi
Comment (headtr1ck) [reply src/netCDF4/__init__.pyi:1]: It was auto generated using an offline tool, but due to the cdefs this is not working reliable and lots of manual changes were necessary.

Future PRs that change the api will have to adopt this file as well to reflect the changes.
But given, that the api has not changed much in the last years I don't think that is going to be a large issue.
The stubtest in the CI should hopefully fail if any PRs forget to change this. But anyway good to keep an eye out as maintainer ;)

I will volunteer to help out if problems arise, feel free to tag me.

File: examples/tutorial.py
Comment (headtr1ck) [reply examples/tutorial.py:32]: Because the CI now runs mypy (a static type checker) on all examples and tests to find any inconsistencies.
This means that the typing of these files should also be correct.

In this instance, the level variable is a `Dimension` here but was later overwritten with a `Variable`, so the type checker raises an error.

The same happens below with `f`, which sometimes was a file handle, and other times a `Dataset`...

**Comment by headtr1ck** [comment]
@jswhit any reason,  `__has_parallel_support__ ` and `__has_ncfilter__` were not imported at module level (only available in `netCDF4._netCDF4`?
I imported them now, but can undo that.

**Comment by RandallPittmanOrSt** [comment]
Ugh, I wish I'd found this PR sooner. I've been working on the same thing but as a [stubs package](https://github.com/RandallPittmanOrSt/netcdf4-python-stubs). Looks like I have some things to learn from your PR, like returning `Any` from `__getitem__` rather than a `Union`.

One fun side-quest I took was writing a [script](https://github.com/RandallPittmanOrSt/netcdf4-python-stubs/blob/master/netCDF4_stubs_merge_docstrings.py) with libcst to enable users to import the docstrings into the stub file.

**Comment by headtr1ck** [comment]
> Ugh, I wish I'd found this PR sooner. I've been working on the same thing but as a [stubs package](https://github.com/RandallPittmanOrSt/netcdf4-python-stubs). Looks like I have some things to learn from your PR, like returning `Any` from `__getitem__` rather than a `Union`.

That's very unfortunate, bad timing, haha. After so many years to start this project at the same time.
It seems that you also have some good ideas of better supporting numpy arrays with dtypes.
Maybe we can merge these things somehow?

> One fun side-quest I took was writing a [script](https://github.com/RandallPittmanOrSt/netcdf4-python-stubs/blob/master/netCDF4_stubs_merge_docstrings.py) with libcst to enable users to import the docstrings into the stub file. 

I don't really know what the advantages are of having the docstrings in the stub files. Do you have an example?

**Comment by RandallPittmanOrSt** [comment]
@headtr1ck 

> Maybe we can merge these things somehow?

Yes. I have gone through our two stubs file side-by-side on paper and will make a few comments here today. I found some fairly minor errors/omissions, I have some thoughts on making `Variable` generic, and I have some thoughts on some of the `TypeAlias`es.

> I don't really know what the advantages are of having the docstrings in the stub files. Do you have an example?

Primarily pop-up help in IDEs, e.g. VSCode. Of course one can use `help()` in a REPL but in my mind having both the types and the docstring available in the IDE is nice, especially for obtuse methods you may not come across frequently. Perhaps "discoverability" is the buzzword. In any case, I decided that integrating the docstrings into the stubs should be done programmatically (keeping stubs up-to-date with code is enough of a chore without wading through all the docstrings). I tried making it a build step for my stubs package but it depends on which `netCDF4` version the user has and that was too confusing. So I decided just to make it a script that came with the stubs so users could run it at-will if they wanted the docstrings.

**Comment by headtr1ck** [comment]
I think having the docstrings in two places sounds like a bad idea. This would require some CI checks of they are the same etc and would make future PRs complicated.

If the current docstrings are not shown in IDEs (I haven't checked that) then one could argue to simply move them over to the stub files all together, but probably in a future PR.

I guess it is up to the maintainers to decide such things.

**Comment by RandallPittmanOrSt** [comment]
> I think having the docstrings in two places sounds like a bad idea. This would require some CI checks to test if they are the same etc and would make future PRs complicated.

I agree completely that they should not be in two places in the netCDF4-python repo. That's why I settled on the idea that the script should be run by users post-install, and only if desired. Even that does present the risk that a user might not re-run the script on updating netCDF4 and have outdated docstrings in their IDE, but that seems fairly low-risk for the benefit (IMO).
 
> If the current docstrings are not shown in IDEs (I haven't checked that) then one could argue to simply move them over to the stub files all together, but probably in a future PR.
> 
> I guess it is up to the maintainers to decide such things.

No, I don't think so, because then the docstrings would be lost from the runtime and, therefore, the REPL. As far as I can tell, the docstrings do belong in the source.

I think the only ideal solution if there was one (to this admittedly niche problem) would be to have a 100% pure-Python API in a `.py` file. And then the annotations could go in there too.

**Comment by headtr1ck** [comment]
I would say, we focus on typing first in this PR and then open a separate issue for the docstrings.
Maybe there is a way to copy over the docstrings when building wheels, so at least there is only a single source?

**Comment by RandallPittmanOrSt** [comment]
# Types

I'd like to address the question of `Variable` types and making `Variable` generic in its
own comment.

I think there are some questions we need to try to answer:

1. What is the benefit of making `Variable` generic?
   - We probably don't want to use the type in `__getitem__()` because we don't know if the
     return type will be an array or a scalar value--it depends on masking configuration
     of the `Dataset` and dimensions of the array and the key.
   - We probably don't want to use the type in `getValue()` because it could return either
     a scalar or a masked array.
   - It doesn't make sense to use the type in `__setitem__` because `__setitem__` is very
     generous with what types it accepts.
   - If we know the type then we can (at least it some/most cases?) annotate the
     `.datatype` and `.dtype` properties... with the caveat that propertly annotating
     `@overload` on properties is a bit of an unknown (see
     [here](https://github.com/python/mypy/issues/9937) and
     [here](https://github.com/microsoft/pyright/issues/3071))
2. What are allowable types for `Variable` (What `bound` for `VarT = TypeVar("VarT", bound=???)`
   used for `class Variable(Generic[VarT]): ...`, if anything)?
   - `CompoundType`, `VLType`, and `EnumType`  are not valid because only an _instance_ of
     any of those types can be the type of a `Variable`.
   - Does it make more sense to use `str` or `np.str_`? Or neither, since it's stored as a
     `VLType` in the file? `str` is the type of returned scalars and class the `VLType`
     uses.
   - To me, (if we are specifying `bound=`) it does _not_ make sense to allow any type
     literals (`"f4"`, `"S1"`, etc.) as those are not types in and of themselves.

I did some research and got this table of results:

| `createVariable` `datatype` parameter                                                                        | `Variable.datatype`                                        | `Variable.dtype`                                    | Notes                                                                                            |
|--------------------------------------------------------------------------------------------------------------|------------------------------------------------------------|-----------------------------------------------------|--------------------------------------------------------------------------------------------------|
| Literal for a scalar numeric (`"i4"`) _except complex types_                                                 | `np.dtype` of that type                                    | `np.dtype` of that type                             | Cannot infer statically without `@overload` for every single literal of that type                |
| Literal `"c8"` or `"c16"`                                                                                    | `CompoundType` instance                                    | `np.dtype('complex64')` or `np.dtype('complex128')` | Cannot infer statically without `@overload` for every single literal of that type                |
| numpy scalar numeric type (`np.int8`) or scalar numeric dtype (`np.dtype(np.uint16)`) _except complex types_ | `np.dtype` of that type                                    | `np.dtype` of that type                             | Can infer statically with a generic `@overload`                                                  |
| numpy complex scalar numeric type (`np.complex64`) or scalar numeric dtype (`np.dtype(np.complex64)`)        | `CompoundType` instance                                    | `np.dtype('complex64')` or `np.dtype('complex128')` | Can infer statically with a generic `@overload`. `datatype` overload must be separate            |
| `int`                                                                                                        | `np.dtype(int64)`                                          | `np.dtype('int64')`                                 | Could add an `@overload` for this. Is `np.dtype(int) == np.dtype('int64')` on all platforms?     |
| `float`                                                                                                      | `np.dtype(float64)`                                        | `np.dtype('float64')`                               | Could add an `@overload` for this. Is `np.dtype(float) == np.dtype('float64')` on all platforms? |
| `str` or `np.str_` or `dtype(str)` or `dtype(np.str_)`                                                       | `VLType` instance with `.name is None` and `.dtype == str` | `str`                                               | Could be provided by a specific `@overload`. I think the `Variable` type should be `str`.        |
| `"S1"`                                                                                                       | `np.dtype('S1')`                                           | `np.dtype('S1')`                                    | Could maybe use `np.bytes_` for the type??                                                       |
| `"S2"` or any other number, `"U"` or `"U1"` or U with any number, or np.dtype with those literals            | `VLType` instance with `.name is None` and `.dtype == str` | `str`                                               | Undocumented. Don't include in stubs except allowing for in an `@overload` with `Any`            |
| `CompoundType` instance                                                                                      | `CompoundType` instance                                    | structured array `dtype`                            | I don't think this can be annotated                                                              |
| `EnumType` instance                                                                                          | `EnumType` instance                                        | `dtype` of the instance                             | I don't think this can be annotated                                                              |
| `VLType` instance                                                                                            | `VLType` instance                                          | `dtype` of the instance                             | I don't think this can be annotated                                                              |

The following cannot be used as `datatype` parameters:

- `np.bytes_`
- Other `np.float` or `np.complex` precisions.
- `bool` or `np.bool_`
- Strutured `np.dtype` -- e.g. `st1 = np.dtype([("x", np.uint8), ("y", np.float32)])`. A
  `CompoundType` must be created first.

Based on all of the above, I propose the following truncated implementation:
<https://gist.github.com/RandallPittmanOrSt/ce695f0b7d7717493649909d1926314a>

File: src/netCDF4/__init__.pyi
Comment (RandallPittmanOrSt) [inline src/netCDF4/__init__.pyi:19]: In my opinion all `Union`s and `Literal`s should be named singular.

File: src/netCDF4/__init__.pyi
Comment (RandallPittmanOrSt) [inline src/netCDF4/__init__.pyi:8]: Rather than import these (since cftime doesn't provide stubs) we can stub them out as follows:

```python
# date2index, date2num, and num2date are actually provided by cftime and if stubs for
# cftime are completed these should be removed.
def date2index(
    dates: (datetime.datetime | cftime.datetime | Sequence[datetime.datetime | cftime.datetime] | DateTimeArr),
    nctime: Variable,
    calendar: CfCalendar | None = None,
    select: Literal["exact", "before", "after", "nearest"] = "exact",
    has_year_zero: bool | None = None,
) -> int | npt.NDArray[np.int_]: ...
def date2num(
    dates: (datetime.datetime | cftime.datetime | Sequence[datetime.datetime | cftime.datetime] | DateTimeArr),
    units: str,
    calendar: CfCalendar | None = None,
    has_year_zero: bool | None = None,
    longdouble: bool = False,
) -> np.number | npt.NDArray[np.number]: ...
def num2date(
    times: Sequence[int | float | np.number] | npt.NDArray[np.number],
    units: str,
    calendar: CfCalendar = "standard",
    only_use_cftime_datetimes: bool = True,
    only_use_python_datetimes: bool = False,
    has_year_zero: bool | None = None,
) -> datetime.datetime | DateTimeArr: ...
```

**Review by RandallPittmanOrSt — CHANGES_REQUESTED** [review]
I recognize that I have included a ton of stuff in this review. I've only reviewed the stubs so I don't know if/how any changes will affect tests etc. I'm going to be offline for the next few weeks so feel free to use or ignore whatever you want from this review.

**Comment by RandallPittmanOrSt** [comment]
I will add, I didn't have any problems with my stub file working under Python 3.8. I did have to get `Buffer`, `Self`, and `TypeAlias` from `typing_extensions`, but otherwise following the docs: <https://typing.readthedocs.io/en/latest/source/stubs.html>

**Comment by jswhit** [comment]
> @jswhit any reason, `__has_parallel_support__ ` and `__has_ncfilter__` were not imported at `netCDF4` level (only available in `netCDF4._netCDF4`? I imported them now, but can undo that.

No reason, just an oversight on my part. OK to fix.

File: src/netCDF4/__init__.pyi
Comment (headtr1ck) [reply src/netCDF4/__init__.pyi:19]: That applies to the variables/arguments but not to the types themselfs?

File: src/netCDF4/__init__.pyi
Comment (headtr1ck) [reply src/netCDF4/__init__.pyi:202]: yes, was mainly for documentation since most IDEs show these type hints together with the docstring

File: src/netCDF4/__init__.pyi
Comment (randallpittman) [reply src/netCDF4/__init__.pyi:19]: My point is type names are generally singular. E.g. `int`, `str`

File: src/netCDF4/__init__.pyi
Comment (headtr1ck) [reply src/netCDF4/__init__.pyi:414]: well, thats because you can change them in runtime (Not saying that you should). So classical attributes are correct here.

Comment (headtr1ck) [reply src/netCDF4/__init__.pyi:414]: Just noticed that the docstring says:
> The instance variables `dtype` and `name` should not be modified by the user.

Maybe that should be ensured in the code? haha

File: src/netCDF4/__init__.pyi
Comment (headtr1ck) [reply src/netCDF4/__init__.pyi:470]: yeah, I feel like the datamodel of `MFDataset` is not very consistent here.
It feels like `_Variable` should be inheriting from `Variable` or something.

Comment (headtr1ck) [reply src/netCDF4/__init__.pyi:470]: I think I will leave it for now and we will see if anyone ever complains.

File: src/netCDF4/_netCDF4.pyi
Comment (headtr1ck) [reply src/netCDF4/_netCDF4.pyi:19]: because it is not imported in `__init__.py`.

**Comment by headtr1ck** [comment]
I would say I leave it like this. The Variable proposal with all the Descriptors looks quite hacky and is beyond my knowledge.
I propose to merge as is, and @RandallPittmanOrSt can open a follow-up PR to improve things.

**Comment by jswhit** [comment]
[stubtest](https://github.com/Unidata/netcdf4-python/actions/runs/9686631247/job/26732005277?pr=1302) is failing

**Comment by headtr1ck** [comment]
> [stubtest](https://github.com/Unidata/netcdf4-python/actions/runs/9686631247/job/26732005277?pr=1302) is failing

Can you try again please?

**Comment by jswhit** [comment]
> > [stubtest](https://github.com/Unidata/netcdf4-python/actions/runs/9686631247/job/26732005277?pr=1302) is failing
> 
> Can you try again please?

still failing

File: src/netCDF4/_netCDF4.pyx
Comment (headtr1ck) [reply src/netCDF4/_netCDF4.pyx:2116]: just to keep things consistent and keep all type hints in the stub (*.pyi) files.

File: docs/index.html
Comment (headtr1ck) [reply docs/index.html:122]: Ah, my bad. Do I revert this or will it just get overwritten?

Comment (jswhit) [reply docs/index.html:122]: please revert

**Review by jswhit — APPROVED** [review]

**Comment by jswhit** [comment]
waiting for @RandallPittmanOrSt to complete his review

**Review by RandallPittmanOrSt — APPROVED** [review]
Sorry, I didn't mean to leave this hanging. I'm fine with going ahead with this PR as-is.

**Comment by jswhit** [comment]
@headtr1ck what is the procedure for updating this when a new function is added to the API (for example, PR #1348)?

**Comment by headtr1ck** [comment]
> @headtr1ck what is the procedure for updating this when a new function is added to the API (for example, PR #1348)?

New or changed API has to be manually added to the stub files. It should be easier directly after writing the code for it, then what we did now...

The stub test should fail if this gets forgotten, so not much intervention required from reviewers/maintainers.

Speaking of, maybe we should add this to some contributing guidelines?

**Comment by jswhit** [comment]
@headtr1ck good idea about adding a CONTRIBUTING.md file - but that can wait for a separate PR.  Merging now...

**Comment by djhoese** [comment]
Hey @jswhit, just curious when this might be available in a release? From the github UI it looks like it hasn't been released (understandable). I just ran into this where my IDE and mypy got mad at me for typing with netCDF4 classes that it didn't understand.

**Pull request merged**
